lucene/core/src/java/org/apache/lucene/util/bkd/BKDReader.java

```diff
@@ -780,4 +783,43 @@ public int getDocCount() {
   public boolean isLeafNode(int nodeID) {
     return nodeID >= leafNodeOffset;
   }
+
+  protected static class BKDReaderDocIDSetIterator extends DocIdSetIterator {
+
+    int idx;
+    int length;
+    int offset;
+    int[] docIds;
+
+    public BKDReaderDocIDSetIterator(int maxPointsInLeafNode) {
+      this.docIds = new int[maxPointsInLeafNode];
+    }
+
+    @Override
+    public int docID() {
+      if (idx == -1)  {
+        return -1;
+      }
+      return docIds[offset + idx];
+    }
+
+    @Override
+    public int nextDoc() throws IOException {
+      if (idx == length - 1) {
+        return DocIdSetIterator.NO_MORE_DOCS;
+      }
+      idx++;
+      return docIds[offset + idx];
+    }
+
+    @Override
+    public int advance(int target) throws IOException {
+      return slowAdvance(target);
+    }
+
+    @Override
+    public long cost() {
+      return length;
+    }
+  }
```

Comment:
can you add tests for it? e.g. making sure `docID()` actually returns -1 when unpositioned and NO_MORE_DOCS when exhausted (I think the latter is not correct at the moment)

Reply:
You are totally right and indeed we should be testing it. I added the test in the verify method. I call randomly the default implementation or check manually the answer from the iterator.

I changed the iterator to hold the current docID so I do not need to check the index of the array all the time.